### PR TITLE
Noam/streaming

### PR DESCRIPTION
**Pull request by nhoffmann-newrelic**
TODO:
add spans
calc tokens
test everything thoroughly

### CODE REVIEW

File: src/nr_openai_observability/stream_patcher.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    build_completion_error_events,
+    build_stream_completion_events,
+)
+from nr_openai_observability.error_handling_decorator import handle_errors
+from nr_openai_observability.openai_monitoring import monitor
+
+
+async def wrap_async_stream_generator(stream_gen):
+    async for chunk in stream_gen:
+        print("chunk----------------------------")
```
Comment (ykriger-newrelic) [inline src/nr_openai_observability/stream_patcher.py:14]: remove print

Comment (ykriger-newrelic) [reply src/nr_openai_observability/stream_patcher.py:14]: Where is it in use?

File: src/nr_openai_observability/stream_patcher.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        raise ex
+
+    wrapped_result = wrap_stream_generator(result)
+
+    return wrapped_result
+
+
+@handle_errors
+def handle_stream_completed(last_chunk, request, error, response_time, message):
+    # encoding = encoding_for_model(last_chunk.get("model") or last_chunk.get("engine"))
```
Comment (ykriger-newrelic) [inline src/nr_openai_observability/stream_patcher.py:98]: what's that?

File: src/nr_openai_observability/consts.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+EventName = "LlmCompletion"
+MessageEventName = "LlmChatCompletionMessage"
+SummeryEventName = "LlmChatCompletionSummary"
```
Comment (umaannamalai) [inline src/nr_openai_observability/consts.py:3]: SummeryEventName -> SummaryEventName

**Comment by umaannamalai** [comment]
Is there a plan to write tests for all of this functionality?

File: pyproject.toml [hunk 4]
```diff
@@ -12,6 +12,7 @@ keywords = ["newrelic", "observability", "openai", "gpt", "chatGPT", "GPT-4", "m
 python = ">=3.8.1,<4.0.0"
 newrelic-telemetry-sdk = "^0.4.0"
 openai = ">=0.8,<0.30"
+tiktoken = "^0.5.1"
```
Comment (umaannamalai) [inline pyproject.toml:15]: Is this dependency being used? (I may have missed it)

File: src/nr_openai_observability/stream_patcher.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    # num_of_tokens = len(encoding.encode(message.get("content", "")))
+    # print(num_of_tokens)
+    events = None
+    if error:
+        events = build_completion_error_events(request, error)
```
Comment (umaannamalai) [inline src/nr_openai_observability/stream_patcher.py:103]: We may want to add a key-value pair in the error event to indicate if the error was from a streamed message.

File: src/nr_openai_observability/consts.py (hunk 3 above)
Comment (nhoffmann-newrelic) [reply src/nr_openai_observability/consts.py]: fixed

File: src/nr_openai_observability/build_events.py [hunk 6]
```diff
@@ -43,6 +43,43 @@ def _get_numeric_header(name):
     }
 
 
+def build_stream_completion_events(
+    last_chunk, request, response_headers, message, response_time
+):
+    print(last_chunk)
```
[the remark this answers is not in the document]
Comment (nhoffmann-newrelic) [reply src/nr_openai_observability/build_events.py]: removed

File: src/nr_openai_observability/stream_patcher.py (hunk 5 above)
Comment (nhoffmann-newrelic) [reply src/nr_openai_observability/stream_patcher.py]: done

File: src/nr_openai_observability/build_events.py [hunk 7]
```diff
@@ -43,6 +43,43 @@ def _get_numeric_header(name):
...
+        "temperature": request.get("temperature"),
+        "max_tokens": request.get("max_tokens"),
+        "finish_reason": last_chunk.choices[0].finish_reason,
+        "api_type": last_chunk.api_type,
+        "vendor": "openAI",
+        "ingest_source": "PythonSDK",
+        "number_of_messages": len(request.get("messages", [])) + 1,
+        "organization": last_chunk.organization,
+        "api_version": response_headers.get("openai-version"),
+    }
```
[the remark this answers is not in the document]
Comment (nhoffmann-newrelic) [reply src/nr_openai_observability/build_events.py:129]: done

**Review by ykriger-newrelic — APPROVED** [review]

**Pull request merged**